test/Test.hs

```diff
@@ -22,8 +44,413 @@ monadReader2 = assertEqual "should be equal" [(5, 0)] $
       y <- ask
       return (x,y)
 
+monadReader3 :: Assertion
+monadReader3 = assertEqual "should be equal" [5,3] $
+  runReader (observeAllT (plus5 `mplus` mzero `mplus` plus3)) (0 :: Int)
+  where
+    plus5 = local (5+) ask
+    plus3 = local (3+) ask
+
+nats, odds, oddsOrTwo,
+  oddsOrTwoUnfair, oddsOrTwoFair,
+  odds5down :: Monad m => LogicT m Integer
+
+#if MIN_VERSION_base(4,8,0)
+nats = pure 0 `mplus` ((1 +) <$> nats)
+#else
+nats = return 0 `mplus` liftM (1 +) nats
```

Comment:
Let's use `liftM` version unconditionally. It's not worth CPP.

Reply:
I was hoping to test the applicative instance definition.  Maybe if I add a comment indicating that is the intent and why it isn't done for GHC7?

Reply:
To be honest, I think the benefit is almost negligible, but I won't block it.
(I've just updated `Control.Monad.Logic` so that `return = pure`)